synapse/handlers/saml_handler.py

```diff
@@ -174,26 +177,56 @@ def handle_redirect_request(self, client_redirect_url):
                     return registered_user_id
 
             # figure out a new mxid for this user
-            base_mxid_localpart = self._mxid_mapper(mxid_source)
+            for i in range(1000):
+                # Use the provider's custom handler if available
+                if self._mapping_provider:
+                    localpart = self._mapping_provider.mxid_source_to_mixd_localpart(
+                        mxid_source, i
```

Comment:
to simplify the interface, we could probably just have the mapping provider generate the base `localpart` and then do the addition of a suffix here?

Reply:
Oh, you're talking about adding the number at the end to dedupe, right? I thought about this, but then thought that providers may want to have their own way of deduping.

Do you think that's reasonable or will everyone be happy with `@john.smith1:example.com`, `@john.smith2:example.com`, ...?

Reply:
it's a fair point. ok, shelve this for a minute and let's see how things look with the other changes.